### PR TITLE
feat(heartbeat): adds periodic requests for health checks

### PR DESCRIPTION
**Pull request by fcanela**
## Summary by CodeRabbit

- **New Features**
  - Added heartbeat functionality to monitor and signal service status.
  - Introduced new configuration options for heartbeat URL and interval.

- **Bug Fixes**
  - Improved logging for heartbeat events.

- **Tests**
  - Added comprehensive test cases for heartbeat service, including intervals, server responses, and event triggers.

- **Chores**
  - Updated dependencies to latest versions.
  - Modified `.env.example` to include new heartbeat configuration variables.

### CODE REVIEW

File: src/services/heartbeat.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+  if (interval) clearInterval(interval);
+  interval = setInterval(() => {
+    fetch(HEARTBEAT_URL)
+      .then((res) => {
+        if (!res.ok) throw new Error("Status code != 2XX");
+        emitter.emit(BotEventNames.HEARTBEAT_SENT);
+      })
+      .catch(() => emitter.emit(BotEventNames.HEARTBEAT_FAILED));
+  }, HEARTBEAT_INTERVAL);
```
[the remark this answers is not in the document]
Comment (fcanela) [reply src/services/heartbeat.ts:31]: The emitted event is received by a logging function, which prints an appropriate error. Doing as suggested would result in a duplicated error line.

File: src/utils/logger.ts [hunk 2]
```diff
@@ -75,4 +76,20 @@ export const configurableInitialize = (deps: ConfigurableInitializeDeps) => {
...
+    logger.info("Perioc hearbeats ready");
+  });
+
+  emitter.on(event.HEARTBEAT_SENT, () => {
+    logger.info(`Perioc hearbeat sent to ${env.HEARTBEAT_URL}`);
+  });
+
+  emitter.on(event.HEARTBEAT_FAILED, () => {
+    logger.error(`Perioc hearbeat to ${env.HEARTBEAT_URL} failed`);
+  });
```
[the remark this answers is not in the document]
Comment (fcanela) [reply src/utils/logger.ts:94]: Good catch, my mechanical friend!

File: src/services/heartbeat.test.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
+import http from "node:http";
+import { expect, resolveOnEvent } from "../utils/tests-setup";
+import { configurableInitialize, interval } from "./heartbeat";
+import EventEmitter from "node:events";
+import { AddressInfo } from "node:net";
+import { Env } from "../utils/env";
+import { BotEventNames } from "../bot-events";
+import { waitFor } from "../bot";
+
+const createFakeServer = async (statusCode: number, onQuery?: Function) => {
```
[the remark this answers is not in the document]
Comment (fcanela) [reply src/services/heartbeat.test.ts:10]: I agree

File: src/services/heartbeat.test.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      } as Env;
+
+      const transportReadyEventPromise = resolveOnEvent(BotEventNames.HEARTBEAT_READY, emitter);
+
+      fn({ emitter, env });
+
+      await waitFor(expectedCalls * interval + interval * 0.5);
+      server.close();
+
+      expect(httpCalls).to.equal(expectedCalls, "web server received less calls than expected");
```
Comment (salgozino) [inline src/services/heartbeat.test.ts:54]: (Ignorable)
Here could be less or more calls than expected, right? The error text should be "web server received a different amount of calls than expected" or something like that.
The same for the line below.

**Review by salgozino — APPROVED** [review]
To me looks really good.
Proceda!

File: src/services/heartbeat.test.ts (hunk 4 above)
Comment (fcanela) [reply src/services/heartbeat.test.ts:54]: @salgozino you are completely right, thanks for catching it! fixing it...

Comment (fcanela) [reply src/services/heartbeat.test.ts]: Fixed! Thanks @salgozino

File: src/services/heartbeat.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+export const configurableInitialize = (deps: ConfigurableInitializeDeps) => {
+  const { env, emitter } = deps;
+
+  const { HEARTBEAT_URL, HEARTBEAT_INTERVAL } = env;
+  if (!HEARTBEAT_URL || !HEARTBEAT_INTERVAL) {
+    emitter.emit(BotEventNames.HEARTBEAT_CONFIGURATION_MISSING);
+    return;
+  }
+
+  if (interval) clearInterval(interval);
```
Comment (wadader) [inline src/services/heartbeat.ts:23]: When does this line come into play?

**Review by wadader — APPROVED** [review]
Looks great! 
All I have are queries for my own curiosity.

In addition to the one comment:

Any service in particular zodiac-bot is expecting to heartbeat?

File: src/services/heartbeat.ts (hunk 5 above)
Comment (fcanela) [reply src/services/heartbeat.ts:23]: @wadader If the service has been already configured and by mistake it is called twice or more times, we remove any previous periodic request (modeled with intervals) before setting the new one. This helps avoiding floods while preserving the ability to call the function multiple times in tests / reconfigure it.

**Comment by fcanela** [comment]
> Any service in particular zodiac-bot is expecting to heartbeat?

BetterStack uptime monitoring

**Review by jaybuidl — APPROVED** [review]

**Pull request merged**